### ISSUE TITLE
Unwanted side effects on saving a XAML with auto-formatting enabled

### ISSUE THREAD

**Issue by drewello** [Xavalon/XamlStyler#98]
Using VS 2017 with XAML Styler 3.0 and Configuration "Format XAML on save" = TRUE

- all collapsed code parts are expanded
- VS Bookmarks are deleted

on every saving of the xaml.

**Comment by NicoVermeir** [comment]
these are side effects of how XAML Styler works. we load your XAML into memory, format it, clear your file completely and write the formatted XAML, causing bookmarks to disappear and opening collapsed parts or regions.

I'm not sure how to tackle this problem...

**Comment by jongleur1983** [comment]
If that's the case it's more difficult to solve this issue, but it should not be impossible.
Without diving into the existing code base yet I see two possible ways to at least partly solve the issue:

One way would be to style the xaml code as it's done now (and as you say, @NicoVermeir), but without writing the result directly to disk overwriting the original file, but instead calculating a diff file between both and applying that diff only.
As there cannot be a conflict (only one side is changed), it should work like diff resolution works with GIT or other VCS, usually keeping bookmarks and expansion state in place where it's not directly affected.

The other way would need a higher coupling to Visual Studio, and I'm not sure about the implications: Keep track of metadata like expansion state and bookmarks and manipulate these accordingly.

**Comment by IanSavchenko** [comment]
I investigated the problem and it didn't reproduce in console mode which does full file rewrite. All the bookmarks and folding are preserved. 

In extension mode, there was an issue when ReplaceText method was called with an empty flag argument. 
Passing value vsEPReplaceTextKeepMarkers helped. Tested on both VS2015 and VS2017. Formatting is smooth and fast both "on save" and from the context menu. Submitting a pull request #119. 

Docs: 
https://docs.microsoft.com/en-us/dotnet/api/envdte.vsepreplacetextoptions.vsepreplacetextkeepmarkers?view=visualstudiosdk-2017
Where I found the value for the flag:
https://github.com/jjensen/solutionbuildenvironment/blob/e1fc5860efdfb1d0e25e3ed70b4bda2cb6edbfee/dte.tlh#L2491
